psutil/arch/windows/process_info.c

```diff
@@ -204,118 +190,346 @@ handlep_is_running(HANDLE hProcess) {
         }
+
+        cmdlineData[ProcBlock.CommandLine.Length / sizeof(WCHAR)] = '\0';
+    } else if (param == PARAMETER_CWD) {
+        // read current working directory
+        cwdData = malloc(ProcBlock.CurrentDirectoryPath.Length + 2);
+        if (!cwdData) {
+            PyErr_NoMemory();
+            goto error;
+        }
+
+        if(!ReadProcessMemory(hProcess, ProcBlock.CurrentDirectoryPath.Buffer, cwdData, ProcBlock.CurrentDirectoryPath.Length, &sRead)) {
+            handle_ReadProcessMemory_error();
+            goto error;
+        }
+
+        cwdData[ProcBlock.CurrentDirectoryPath.Length / sizeof(WCHAR)] = '\0';
+    } else if (param == PARAMETER_ENVIRON) {
+        pEnvStr = ProcBlock.env;
+    }
+
     }
 
-    if (szArglist != NULL)
+    if (param == PARAMETER_CMDLINE) {
+        szArglist = CommandLineToArgvW(cmdlineData, &nArgs);
+
+        if (szArglist == NULL) {
+            PyErr_SetFromWindowsErr(0);
+            goto error;
+        }
+
+        ret = PyList_New(nArgs);
+        if (ret == NULL)
+            goto error;
+
+        // parse the command line using Win32 API
+        {
+            int i;
+            PyObject *py_item ;
+            for (i = 0; i < nArgs; i++) {
+                py_item = PyUnicode_FromWideChar(szArglist[i], wcslen(szArglist[i]));
+                if (py_item == NULL)
+                    goto error;
+                PyList_SET_ITEM(ret, i, py_item);
+            }
+        }
+
         LocalFree(szArglist);
-    free(commandLineContents);
-    CloseHandle(hProcess);
-    return py_retlist;
+        szArglist = NULL;
+    } else if (param == PARAMETER_CWD) {
+        ret = PyUnicode_FromWideChar(cwdData, wcslen(cwdData));
+        if (ret == NULL)
+            goto error;
+
+        free(cwdData);
+    } else if (param == PARAMETER_ENVIRON) {
+        // figure out the size of the env var block
+        MEMORY_BASIC_INFORMATION info;
+        SIZE_T envOffset;
+        if(VirtualQueryEx(hProcess, pEnvStr, &info, sizeof(info))==0) {
+            PyErr_SetFromWindowsErr(0);
+            goto error;
+        }
+
+        envData = malloc(info.RegionSize);
+        if (envData == NULL) {
+            PyErr_NoMemory();
+            goto error;
+        }
+        memset(envData, 0, info.RegionSize);
+
+        // Reading the whole reading seems to work more reliably than reading
+        // just from the environment pointer.
+        ReadProcessMemory(hProcess, info.BaseAddress, envData, info.RegionSize, &sRead);
+        envOffset = (char*)pEnvStr - (char*)info.BaseAddress;
+
+        if(sRead < envOffset) {
+            handle_ReadProcessMemory_error();
+            goto error;
+        }
+
+        ret = PyUnicode_FromWideChar((LPWSTR)((char*)envData + envOffset), (sRead - envOffset) / 2);
+        if (ret == NULL)
+            goto error;
+
+        free(envData);
+        envData = NULL;
+    }
+
+    if (hProcess != NULL)
+        CloseHandle(hProcess);
+
+    return ret;
```

Comment:
Okay, but that would mean doing the parsing in C.  It's doable but much more error-prone IMHO.